spark/src/main/scala/io/delta/sharing/spark/filters/JsonPredicates.scala

```diff
@@ -0,0 +1,346 @@
+}
+
+case class GreaterThanOp(children: Seq[LeafOp]) extends NonLeafOp with BinaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any =
+    !EvalHelper.lessThan(children, ctx) && !EvalHelper.equal(children, ctx)
+}
+
+case class GreaterThanOrEqualOp(children: Seq[LeafOp]) extends NonLeafOp with BinaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = !EvalHelper.lessThan(children, ctx)
+}
+
+/**
+ * The following ops represent boolean logic related operations.
+ * The (logically) supported operations are:
+ *   "&&", "||", "!"
+ */
+
+case class AndOp(val children: Seq[NonLeafOp]) extends NonLeafOp with NaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = {
+    children.forall(c => c.evalExpectBoolean(ctx))
+  }
+}
+
+case class OrOp(val children: Seq[NonLeafOp]) extends NonLeafOp with NaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = {
+    children.exists(c => c.evalExpectBoolean(ctx))
+  }
+}
+
+case class NotOp(val children: Seq[NonLeafOp]) extends NonLeafOp with UnaryOp {
+  override def validate(): Unit = validateChildren(children)
+
+  override def eval(ctx: EvalContext): Any = !children(0).evalExpectBoolean(ctx)
+}
+
+// A helper for evaluating opertions.
+object EvalHelper {
+
+  // Implements "equal" between two leaf operations.
+  def equal(children: Seq[LeafOp], ctx: EvalContext): Boolean = {
+    val (leftVal, leftType) = children(0).evalExpectValueAndType(ctx)
+    val (rightVal, rightType) = children(1).evalExpectValueAndType(ctx)
+    // If the types don't match, it implies a malformed predicate tree.
+    // We simply throw an exception, which will cause filtering to be skipped.
+    if (leftType != rightType) {
+      throw new IllegalArgumentException(
+        "Type mismatch: " + leftType + " vs " + rightType + " for " +
+         children(0) + " and " + children(1)
+      )
+    }
+    // Handle nulls.
+    if (leftVal == null && rightVal == null) {
+      return true
+    } else if (leftVal == null || rightVal == null) {
+      return false
+    }
+    // Perform data type conversion and match.
+    // TODO(abhijit): For literal operations, we can optimize evaluation by caching
+    //                this conversion.
+    leftType match {
+      case OpDataTypes.BoolType => leftVal.toBoolean == rightVal.toBoolean
+      case OpDataTypes.IntType => leftVal.toInt == rightVal.toInt
+      case OpDataTypes.LongType => leftVal.toLong == rightVal.toLong
+      case OpDataTypes.StringType => leftVal == rightVal
+      case OpDataTypes.DateType =>
+        java.sql.Date.valueOf(leftVal).equals(java.sql.Date.valueOf(rightVal))
+      case _ =>
+        throw new IllegalArgumentException("Unsupported type: " + leftType)
+    }
+  }
```

Comment:
I don't like this type handling.
Any thoughts on simplifying this structure are welcome?

Reply:
If you're enforcing the same type already, since `leftVal` and `rightVal` are both string, you can just compare them directly. It would have been nicer to pass the actual types from the bottom though instead of converting from String to their native types.